datafusion/src/scalar.rs

```diff
@@ -1095,6 +1105,78 @@ impl ScalarValue {
         }
     }
 
+
+    /// an unsafe version of `eq_array` optimised for the situation
+    /// where we know in advance that the array has no nulls
+    #[inline]
+    pub fn eq_array_no_nulls(&self, array: &ArrayRef, index:usize) -> bool {
+        if let DataType::Dictionary(key_type, _) = array.data_type() {
+            return self.eq_array_dictionary(array, index, key_type);
```

Comment:
I think this should perhaps also be `self.eq_array_dictionary_no_nulls` and have a corresponding dictionary special case.